app/components/UserModal.tsx

```diff
@@ -0,0 +1,197 @@
+import { useState } from "react";
+import Alert from "react-bootstrap/Alert";
+import Button from "react-bootstrap/Button";
+import Form from "react-bootstrap/Form";
+import Modal from "react-bootstrap/Modal";
+import Spinner from "react-bootstrap/Spinner";
+import { useForm, SubmitHandler, Controller } from "react-hook-form";
+import { User, UserAPIError } from "@/types/users";
+import { useToken } from "@/utils/auth";
+
+interface UserModalProps {
+  /**
+   * A callback fired when either the modal backdrop is clicked, or the
+   * escape key is pressed
+   */
+  onClose: () => void;
+  /**
+   * An async callback fired after the user form is successfully submitted
+   */
+  onSubmitCallback: (user: User) => Promise<void>;
+  /**
+   * If the modal should be visible or hidden
+   */
+  show: boolean;
+  /**
+   * An optional User object to be edited by this component. If not defined, a new user
+   * will be created on submit
+   */
+  user?: User;
+}
+
+type UserInputs = {
+  name: string;
+  email: string;
+  app_metadata: { roles: string[] };
+};
+
+const USER_DEFAULTS = {
+  name: undefined,
+  email: undefined,
+  app_metadata: { roles: ["readonly"] },
+};
+
+/**
+ * Modal form component used for creating or editing a user
+ */
+export default function UserModal({
+  onClose,
+  onSubmitCallback,
+  show,
+  user,
+}: UserModalProps) {
+  const [isSubmitting, setIsSubmitting] = useState(false);
+  const [errorMessage, setErrorMessage] = useState<string | null>(null);
+  const token = useToken();
+  const {
+    register,
+    handleSubmit,
+    control,
+    formState: { errors, isDirty },
+  } = useForm<UserInputs>({
+    defaultValues: user
+      ? { name: user.name, email: user.email, app_metadata: user.app_metadata }
+      : USER_DEFAULTS,
+  });
+
+  const onSubmit: SubmitHandler<UserInputs> = async (data) => {
+    setErrorMessage(null);
+    setIsSubmitting(true);
+    const url = user
+      ? `${
+          process.env.NEXT_PUBLIC_CR3_API_DOMAIN
+        }/user/update_user/${encodeURIComponent(user.user_id)}`
+      : `${process.env.NEXT_PUBLIC_CR3_API_DOMAIN}/user/create_user`;
+
+    const method = user ? "PUT" : "POST";
+    try {
+      const response = await fetch(url, {
+        headers: {
+          Authorization: `Bearer ${token}`,
+          "Content-Type": "application/json",
+        },
+        method,
+        body: JSON.stringify(data),
+      });
+      const responseJson: User | UserAPIError = await response.json();
+      if ("user_id" in responseJson) {
+        await onSubmitCallback(responseJson);
+      } else if ("message" in responseJson) {
+        setErrorMessage(responseJson.message);
```

Comment:
why are we checking for the word "message" here?

Reply:
this is a hack for checking if we have an ok response or an error response, since only an error response will have a `message` property. I wrote this before I fixed the problem of the API returning a `200` status with errors.  I reworked this to use the `response.status` code 🙏